### PR TITLE
v4: Increase Bulk Download Speeds

### PR DESCRIPTION
**Pull request by Abdelrahman-Sherif**
Vastly increased download speeds by downloading tiles in bulk.

Earlier, downloading 200 tiles (which is a very small area, around 2 MB) would take 30 seconds on a good internet.

Reason was, the code was awaiting each tile to download and save, before moving on to the next one.
The changes proposed download the tiles in bulk, which made my downloads more than 10x faster.

I would prefer to have more frequent progress responses, but this would require a code re-architecture in some sense, which I unfortunately don't have the time for at the moment, but would love to do so, in the future.

For the moment, the current solution works perfectly for me.

### CODE REVIEW

**Comment by JaffaKetchup** [comment]
Hi there, and thanks for the contribution!

This does look like like it would boost download speed considerably, and it is therefore of great interest to me, and, I'm sure, to other people.

I'm not sure about this line you said though:
> I would prefer to have more frequent progress responses

I haven't tested the code (I can't do it right now), but purely by looking at it, I'm not sure why it would slow down the report of `DownloadProgress`? Currently, the `DownloadProgress` is returned/yielded every tile, would this not happen anymore? Please explain in more detail, it would help a lot :).

Also, I'm not sure if this is an issue, but this code looks like  it start multiple asynchronous download threads. This is a large issue if it is one, because many servers that allow bulk downloading have a clause to say "No more than 2 download threads", such as https://operations.osmfoundation.org/policies/tiles/.
Therefore, when (probably 'when' instead of 'if') I merge this, I will probably have to make this opt-in functionality. Alternatively, I might say "Not my responsibility for people to use tile servers right" and just merge as-is.

So, TL;DR: Please explain a bit more about `DownloadProgress`, and I'll try to have this merged soon!

Thanks once more!

**Comment by Abdelrahman-Sherif** [comment]
Hey JaffaKetchup, thanks for the prompt response!

What I intended to say was, even though this helped me with my download time issue, another (issue?) rose which is the frequency of the DownloadProgress events returned to the caller. 

Earlier, progress events were reported promptly every single tile. However, now I await the bulk download method, then yield the DownloadProgress with the returned bulk data. 
Currently, the count parameter for the bulk download method is tiles.length/4, hence, the caller only gets an event with the DownloadProgress every 25%.

I could not find any solution to this frequency issue without changing major parts of the code, maybe you have something else on mind? I understand that this might be a problem. 
I am currently in the middle of a project with a client and would love to contribute to this package after.

Also, I did not know about the threads limit for some APIs, Thanks for mentioning it! I checked with the API I am using (Swisstopo) and they do not have a limit in place. 
Surely, as you mentioned, this would be an issue when such an option is live. Possibly adding a `required int downloadThreads` parameter to the download methods would help?

This is my first ever contribution, please feel free to ask me anything/request more details if I missed something.

Thanks for this great package, all the best!

**Comment by JaffaKetchup** [comment]
Hi there again.
Sure I would like to collaborate when you're ready!

Just one more question (I think :)): let's say there are 100 tiles, are we are using `tiles.length/4`, how many threads would be simultaneously occuring? 25 or 4?

Thanks.

**Comment by Abdelrahman-Sherif** [comment]
Using the example you mentioned, that would mean 25 simultaneous requests. I tried Googling whether that would be bad or not, but I'm not getting any results. 

I tested it on my Android emulator and physical phone, as well as my personal iPhone and it all went well (with a big map). I understand that with bigger maps this would mean much more threads. We can either clamp it to a value or leave it for developers to pick.
I am not an expert on this topic, so please correct me if I'm wrong :)

I was also thinking of using a slightly better method where we try to maximize utilization. 
Meaning, referencing the example you provided, we currently wait for all 25 threads to complete before making a new bulk download request. It would be much better if we can keep the 25 threads always busy by promptly creating a new download request once a single tile has completed downloading. Something like a bulk queuing system, but instead of picking 25, wait for them to complete, then take another 25. We would take the first 25, then once one completes, we fill its position with another one from the queue.
I believe this would improve performance even further, what do you think?

As I mentioned previously, this is my first contribution. I did not know that every single commit I make would show here. I'll try to keep it tidy from now on :)

**Comment by JaffaKetchup** [comment]
Ok thanks for the confirmation.

I think 25 download threads is a little too many for 100 tiles, but I also want to leave it up to the developer.
So I'll probably make a property `double maxThreads`. Then if they put in '5', for example, the maths to get the number of batches (currently hard-coded to '4') will be `(tiles.length/maxThreads).ceil()`, I think? Not fully sure about the `.ceil()` (rounding up) part, but that seems to make sense? So `count` would be `tiles.length/((tiles.length/maxThreads).ceil())`?

I like the idea about queueing tiles so threads are always utilized, but I'm not sure how to go about coding this. And I'm also not sure if this would affect the download progress reports further?

Don't worry about the messy commits, I can use 'Squash & Merge' to merge all your commits into 1 😄.

**Comment by Abdelrahman-Sherif** [comment]
Hey again, great news!

If you check my latest commit, I believe we have solved all of the mentioned issues :)

Using the Queue package you mentioned, utilization is always at 100% of the threads the user sets. I have changed the code to make use of this package and seen significant improvements.

I have also added a static variable `static var maxParallel = 20` that is used to set the size of the queue. It is preset to 20 and can be changed to the user's liking.

Finally, I have changed up the code a little, and the DownloadProgress issue is now fixed :D 
Progress is now reported promptly with every tile.

I am really happy with the new changes, please check them out when you can and let me know what you think.
All the best!

**Comment by JaffaKetchup** [comment]
Thanks for this great work! I'm going to just change up a few things, mostly just moving some things around, then I'll merge.

**Review by JaffaKetchup — APPROVED** [review]
I've finally gotten round to testing it and the performance increase is substantial, way over 50% reduced download speeds when running the automated tests (on 20 threads)!

- Moved `maxParallelThreads` from `StorageCachingTileProvider` to `parallelThreads` in `DownloadableRegion`
- Tidied up some old code (I've got another commit waiting in the wings to do this and a lot more)
- Renamed some internal methods

I've made these changes, and I'm pretty sure this is good to go now.
Thank you very much for your help, it's greatly appreciated. If there are any last changes you wish to make, please do so now, as I will merge soon.

**Comment by Abdelrahman-Sherif** [comment]
I believe that was everything, really happy with how it's shaping up!

Thank you for your prompt support and this great package, wish you all the best!
